apache/cloudstack
# Dynamic roles improvements

## Description


This PR addresses the below improvements for current Dynamic roles functionality.
- Create a role from any of the existing role, using new parameter roleid in createRole API.
- Import a role with its rules, using a new importRole API.
- New default roles for Read-Only and Support Admin & User.
- No modifications (update or delete) allowed for Default roles.

- Cleanup of old NetApp APIs from role_permissions table.

UI changes PR: https://github.com/apache/cloudstack-primate/pull/353








## Types of changes


## Screenshots (if appropriate):

## How Has This Been Tested?



**Manual Tests using API**

- Create a role using existing role.

http://myms:8080/client/api?command=createRole&response=json&name=Test%20CloneUser01&description=Test%20CloneUser01&roleid=ca9871c2-8ea7-11ea-944e-c2865825b006

{"createroleresponse":{"role":{"id":"c3888c27-66b6-4a39-96ee-9883fef326ba","name":"Test CloneUser01","type":"User","description":"Test CloneUser01"}}}

- Create a role using both role type and role id.

http://myms:8080/client/api?command=createRole&response=json&type=Admin&name=Test%20CloneResourceAdmin&description=Test%20CloneResourceAdmin&roleid=ca984d00-8ea7-11ea-944e-c2865825b006

{"createroleresponse":{"uuidList":[],"errorcode":431,"cserrorcode":9999,"errortext":"Both role type and role ID should not be specified"}}

- Create a role without role type and role id.

http://myms:8080/client/api?command=createRole&response=json&name=Test%20CloneResourceAdmin&description=Test%20CloneResourceAdmin

{"createroleresponse":{"uuidList":[],"errorcode":431,"cserrorcode":9999,"errortext":"Neither role type nor role ID is provided"}}

- Import role

http://myms:8080/client/api?command=importRole&response=json&name=ImportTestUser&type=User&description=Test%20Import%20Role&rules[0].rule=create*&rules[0].permission=allow&rules[0].description=create%20rule&rules[1].rule=list*&rules[1].permission=allow&rules[1].description=listing&forced=true

{"importroleresponse":{"role":{"id":"8306f86d-ab33-48b8-8263-535a0da5be04","name":"ImportTestUser","type":"User","description":"Test Import Role"}}}

- Import role with same name without force

http://myms:8080/client/api?command=importRole&response=json&name=ImportTestUser&type=User&description=Test%20Import1%20Role&rules[0].rule=create*&rules[0].permission=allow&rules[0].description=create%20rule&rules[1].rule=list*&rules[1].permission=allow&rules[1].description=listings

{"importroleresponse":{"uuidList":[],"errorcode":530,"cserrorcode":4250,"errortext":"Role already exists"}}

**Unit Tests**

[INFO] Running org.apache.cloudstack.api.command.test.ImportRoleCmdTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.032 s - in org.apache.cloudstack.api.command.test.ImportRoleCmdTest
[INFO] Running org.apache.cloudstack.api.command.test.CreateRoleCmdTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.018 s - in org.apache.cloudstack.api.command.test.CreateRoleCmdTest

**Smoke Tests**

=== TestName: test_default_role_deletion | Status : SUCCESS ===

=== TestName: test_role_account_acls | Status : SUCCESS ===

=== TestName: test_role_account_acls_multiple_mgmt_servers | Status : SUCCESS ===

=== TestName: test_role_inuse_deletion | Status : SUCCESS ===

_**=== TestName: test_role_lifecycle_clone | Status : SUCCESS ===**_

=== TestName: test_role_lifecycle_create | Status : SUCCESS ===

=== TestName: test_role_lifecycle_delete | Status : SUCCESS ===

_**=== TestName: test_role_lifecycle_import | Status : SUCCESS ===**_

=== TestName: test_role_lifecycle_list | Status : SUCCESS ===

=== TestName: test_role_lifecycle_update | Status : SUCCESS ===

=== TestName: test_role_lifecycle_update_role_inuse | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_concurrent_updates | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_create | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_delete | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_list | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_update | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_update_permission | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_update_permission_negative | Status : SUCCESS ===
